### PR TITLE
WPF upgrade

### PR DESCRIPTION
**Pull request by SadPencil**
Why WPF:

- WinForm has the layout issue which depends on system font and monitor DPI. Not easy to solve (although we can just increase the window size as a workaround).
- `MessageBox` is not a good tool to display an exception message due to missing scrollbars.
- No need to maintain two windows for gpu message and missing component, respectively. One `AdvancedMessageBox` is enough.
- No need to be compactible with Wine anymore since Client is .NET 7. 

Screenshots:
Snipaste_2023-02-19_15-44-10

### CODE REVIEW

**Review by devo1929 — APPROVED** [review]

File: CommandViewModel.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+using System.Windows.Input;
+
+public class CommandViewModel : NotifyPropertyChangedBase
+{
+    private string? text;
+
+    public string Text
+    {
+        get => text; set
```
Comment (Rampastring) [inline CommandViewModel.cs:11]: Put the setter on its own line, looks weird to have it after the semicolon of the getter.

File: CommandViewModel.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+            text = value;
+            NotifyPropertyChanged();
+        }
+    }
+
+    private ICommand? command;
+
+    public ICommand Command
+    {
+        get => command; set
```
Comment (Rampastring) [inline CommandViewModel.cs:22]: Same here as above.

File: Program.cs [hunk 3]
```diff
@@ -120,6 +154,20 @@ private static void RunUGL()
     private static string GetClientProcessPath(string directory, string file)
         => $"{Resources}\\{Binaries}\\{directory}\\{file}";
 
+    private static int? ShowIncompatibleGPUMessage(string[] selections)
+    {
+        return AdvancedMessageBoxHelper.ShowMessageBoxWithSelection(
+            string.Format(
+                "The client has detected an incompatibility between your graphics card and the DirectX11 and OpenGL versions of the CnCNet client.\n\n" +
```
Comment (Rampastring) [inline Program.cs:161]: This line is much longer than the lines below, making the message excessively wide and its layout weird. Split it from the middle, like the line below.

**Review by Rampastring — CHANGES_REQUESTED** [review]

**Comment by SadPencil** [comment]
Snipaste_2023-03-20_12-17-48
Updated

**Comment by Metadorius** [comment]
> Snipaste_2023-03-20_12-17-48 Updated


1) What does this even mean? 
image
2) If there's an incompatibility between PC and DX11 version, why is this an option?
image

**Comment by SadPencil** [comment]
@Metadorius 
> What does this even mean?

This dialog is launched by the dialog test, because it is hard to reproduce the incompatibility just in order to show the prompt. The only difference is the dialog launched by the dialog test will not do anything and has a `(All buttons here won't work)` text.

> If there's an incompatibility between PC and DX11 version, why is this an option?

This PR only migrates the project from WinForm to WPF. No texts or buttons are changed.
Besides, `Alternatively, you can retry launching the DirectX11 version of the client.`

File: RelayCommand.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+    public bool CanExecute(object parameter)
+    {
+        return true;
+    }
+
+    public void Execute(object parameter)
+    {
+        execute(parameter);
+    }
```
Comment (Metadorius) [inline RelayCommand.cs:25]: This begs for expression bodies.

File: Program.cs [hunk 5]
```diff
@@ -75,35 +75,69 @@ private static void Main(string[] args)
...
+                }),
+            },
+
+            new CommandViewModel()
+            {
+                Text = "Exit",
+                Command = new RelayCommand(_ =>
+                {
+                    msgbox.Close();
+                }),
```
Comment (Metadorius) [inline Program.cs:128]: Unnecessary lambda brackets here.

File: Program.cs [hunk 6]
```diff
@@ -231,15 +272,27 @@ private static FileInfo CheckAndRetrieveDotNetHost(Architecture machineArchitect
...
+        bool dialogResult = AdvancedMessageBoxHelper.ShowYesNoMessageBox(
+            string.Format(
+            "The component {0} is missing.\n\n" +
+            "You can download the installer from the following link:\n\n{1}",
+            missingComponent, downloadLink.ToString()), "Component Missing",
+            yesText: "Open link", noText: "Exit");
+        if (dialogResult)
+        {
+            OpenUri(downloadLink);
+        }
```
Comment (Metadorius) [inline Program.cs:295]:
```suggestion

        if (dialogResult)
            OpenUri(downloadLink);
```

File: AdvancedMessageBoxHelper.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+            int i_copy = i;
+            commands.Add(new CommandViewModel()
+            {
+                Text = selections[i],
+                Command = new RelayCommand(_ =>
+                {
+                    msgbox.Result = i_copy;
+                    msgbox.Close();
+                }),
+            });
```
Comment (Metadorius) [inline AdvancedMessageBoxHelper.cs:26]: 1. The purpose of `i_copy` is unobvious for those who don't know about capture issues in C# (I didn't and had to guess and google for example).
2. Naming violation.

```suggestion
            // passing just i to the lambda will not work as
            // intended due to C# variable capture specifics
            int iCaptured = i;
            commands.Add(new CommandViewModel()
            {
                Text = selections[i],
                Command = new RelayCommand(_ =>
                {
                    msgbox.Result = iCaptured;
                    msgbox.Close();
                }),
            });
```

File: NotifyPropertyChangedBase.cs
Comment (Metadorius) [inline NotifyPropertyChangedBase.cs:1]: I recommend to use ReactiveUI.Fody for autogenerating INPC boilerplate.

**Review by Metadorius — CHANGES_REQUESTED** [review]

File: NotifyPropertyChangedBase.cs
Comment (SadPencil) [reply NotifyPropertyChangedBase.cs:1]: This requires distributing a DLL where launcher is supposed single-file contained

**Pull request merged**